src/storage/src/source/postgres/replication.rs

```diff
@@ -0,0 +1,556 @@
+    table_info: BTreeMap<u32, (usize, PostgresTableDesc, Vec<MirScalarExpr>)>,
+    rewind_stream: &Stream<G, RewindRequest>,
+    committed_uppers: impl futures::Stream<Item = Antichain<MzOffset>> + 'static,
+) -> (
+    Collection<G, (usize, Result<Row, SourceReaderError>), Diff>,
+    Stream<G, Infallible>,
+    Stream<G, Rc<anyhow::Error>>,
+    Rc<dyn Any>,
+) {
+    let op_name = format!("ReplicationReader({})", config.id);
+    let mut builder = AsyncOperatorBuilder::new(op_name, scope);
+
+    let slot_reader = u64::cast_from(config.responsible_worker("slot"));
+    let mut rewind_input = builder.new_input(rewind_stream, Exchange::new(move |_| slot_reader));
+    let (mut data_output, data_stream) = builder.new_output();
+    let (_upper_output, upper_stream) = builder.new_output();
+
+    let metrics = PgSourceMetrics::new(&config.base_metrics, config.id);
+    metrics.tables.set(u64::cast_from(table_info.len()));
+
+    let reader_table_info = table_info.clone();
+    let (button, errors) = builder.build_fallible(move |caps| {
+        let table_info = reader_table_info;
+        Box::pin(async move {
+            let (id, worker_id) = (config.id, config.worker_id);
+            let [data_cap, upper_cap]: &mut [_; 2] = caps.try_into().unwrap();
+            let (data_cap, upper_cap) = (data_cap.as_mut().unwrap(), upper_cap.as_mut().unwrap());
+
+            if !config.responsible_for("slot") {
+                return Ok(());
+            }
+
+            let Some(resume_lsn) = resume_upper.into_option() else {
+                return Ok(());
+            };
+            data_cap.downgrade(&resume_lsn);
+            upper_cap.downgrade(&resume_lsn);
+
+            let connection_config = connection
+                .connection
+                .config(&*context.secrets_reader)
+                .await?
+                .replication_timeouts(config.params.pg_replication_timeouts.clone());
+
+            let mut client = connection_config.connect_replication().await?;
+
+            let min_resume_lsn = super::ensure_replication_slot(&client, &connection).await?;
+            assert!(resume_lsn == MzOffset::from(0) || min_resume_lsn <= resume_lsn, "slot compacted past resumption point. resume_lsn={resume_lsn} min_resume_lsn={min_resume_lsn}");
+
+            let mut rewinds = BTreeMap::new();
+            while let Some(event) = rewind_input.next_mut().await {
+                if let AsyncEvent::Data(cap, data) = event {
+                    let cap = cap.retain_for_output(0);
+                    for req in data.drain(..) {
+                        assert!(min_resume_lsn <= req.snapshot_lsn, "slot compacted past snapshot point. snapshot_lsn={} min_resume_lsn={min_resume_lsn}", req.snapshot_lsn);
+                        rewinds.insert(req.oid, (cap.clone(), req));
+                    }
+                }
+            }
+
+            let mut committed_uppers = pin!(committed_uppers);
+
+            loop {
```

Comment:
can you add a brief comment about what this loop retries?

Reply:
yes, added!